### PR TITLE
[Breaking] Validation model update

### PR DESCRIPTION
**Pull request by KentonParton**
# Description
The existing validation structure contains just the expectation details and its results. With these changes we gain a lot more metadata about the validation supporting longer term features such as actions, replay, local client.

This is a breaking change. All existing validations will need to be removed and can be removed with the following OpenSearch query.

```
POST validations/_delete_by_query
{
  "size": 10000, 
  "query": {
    "match_all": {}
  }
}
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have performed a self-review of my own code
- [x] All GitHub workflows have passed
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: backend/app/repositories/base.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def query(self, body: dict[str, Any], *, size: int = 100) -> list[M]:
         ]
 
     def count(self, body: dict[str, Any]) -> int:
-        return self.client.count(index=self.index, body=body)
+        return self.client.count(index=self.index, body=body)["count"]
```
Comment (frankie567) [inline backend/app/repositories/base.py:32]: Oh, good catch 😄

File: backend/app/repositories/validation.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            "query": {
+                "bool": {
+                    "must": [
+                        {"range": {"meta.run_id.run_time": {"gte": f"now-{period}d", "lte": "now"}}},
+                    ]
+                }
+            },
+            "sort": [{"meta.run_id.run_time": "asc"}]
+        }
+        if not dataset_id and not datasource_id:
```
Comment (frankie567) [inline backend/app/repositories/validation.py:29]: We shouldn't raise `HTTPException` in those methods. The database layer should be independent of the API layer. If you need to validate this, do it directly in the endpoint or in a dependency.

File: backend/app/core/runner.py [hunk 3]
```diff
@@ -142,10 +147,10 @@ def validate(self):
                 result["result"]["observed_value_list"] = result["result"].pop("observed_value")
 
             utils.list_to_string_mapper(result)
-            self.identifiers["expectation_id"] = result["expectation_config"]["meta"].pop("expectation_id")
-            result.update(self.identifiers)
+            result["expectation_id"] = result["expectation_config"]["meta"].pop("expectation_id")
 
-        return results
+        print(validation)
```
Comment (frankie567) [inline backend/app/core/runner.py:152]: `print` 🙃

File: backend/app/core/runner.py [hunk 4]
```diff
@@ -244,19 +247,19 @@ def run_dataset_validation(dataset_id: str):
         runner_expectation["meta"] = {"expectation_id": expectation.key}
         runner_expectations.append(runner_expectation)
 
-    results = Runner(
+    validation: Validation = Runner(
```
Comment (frankie567) [inline backend/app/core/runner.py:250]: The method is correctly typed, so you don't need to manually type-hint the result:

```
validation = Runner(
    ...
).validate()
```

File: backend/app/api/api_v1/endpoints/expectation.py [hunk 5]
```diff
@@ -77,11 +81,10 @@ def list_expectations(
     )
 
     if include_history:
-        results = client.search(
-            body=validation.validations_query_body(datasource_id, dataset_id),
-            index=settings.VALIDATION_INDEX,
+        validations: list[Validation] = validation_repository.query_by_filter(
```
Comment (frankie567) [inline backend/app/api/api_v1/endpoints/expectation.py:84]: The method is correctly typed, so you don't need to manually type-hint the result.

File: backend/app/api/api_v1/endpoints/expectation.py [hunk 6]
```diff
@@ -172,16 +176,38 @@ def delete_expectation(
...
 
     for expectation in expectations:
         expectations_as_dict[expectation.key] = expectation
 
-    for v in validations:
-        source = v["_source"]
-        source["run_date"] = utils.string_to_military_time(source["run_date"])
-        expectations_as_dict[source["expectation_id"]].validations.append(source)
+    for validation in validations:
+        run_time = utils.string_to_military_time(validation.meta.run_id.run_time)
```
Comment (frankie567) [inline backend/app/api/api_v1/endpoints/expectation.py:186]: Never came across "military time" before. Does it mean UTC? 😄

File: backend/app/api/api_v1/endpoints/expectation.py [hunk 7]
```diff
@@ -172,16 +176,38 @@ def delete_expectation(
...
+                expectations_as_dict[result.expectation_id].validations.append(result_as_dict)
 
     return list(expectations_as_dict.values())
+
+
+def _table_level_expectation_already_exists(expectation: Expectation, repository: ExpectationRepository):
+    # Duplicate Table level/ result_type="expectation", expectations are removed by GE when validations are run.
+    # Because of this, we want to prevent duplicate table level expectations from being added.
+    if expectation.result_type == c.EXPECTATION:
+        if len(repository.query_by_filter(
```
Comment (frankie567) [inline backend/app/api/api_v1/endpoints/expectation.py:205]: Could be interesting in terms of performance to use `count` here.

What we could do is to factorize the query build into a private method of the repository and use it both in `query_by_filter` and  a new method `count_by_filter`.

**Review by frankie567 — COMMENTED** [review]
That's lot of changes but I think it's for the good: having a model for Validation is very useful and we already see the code is much cleaner like this!

Added some minor comments. The main blocker IMO is the `HTTPException` inside the repository.

File: backend/app/repositories/validation.py (hunk 2 above)
Comment (KentonParton) [reply backend/app/repositories/validation.py]: Agreed, moved it out to the API layer.

File: backend/app/core/runner.py (hunk 3 above)
Comment (KentonParton) [reply backend/app/core/runner.py]: Removed 🙃

File: backend/app/api/api_v1/endpoints/expectation.py (hunk 6 above)
Comment (KentonParton) [reply backend/app/api/api_v1/endpoints/expectation.py]: Yeah, don't know why I did this. It doesn't look like the `string_to_utc_time` method was used to I updated it to use the `string_to_military_time` implementation.

File: backend/app/api/api_v1/endpoints/expectation.py (hunk 7 above)
Comment (KentonParton) [reply backend/app/api/api_v1/endpoints/expectation.py]: I like this. I implemented the filtering logic in `_build_query_filter` which `delete_by_filter` and `count_by_filter` now use.

I think we will want to create some "query builder" in `base` repository if the number of filters grow any more.

**Comment by frankie567** [comment]
LGTM 👍

**Pull request merged**